README.md
[![pulse logo](http://nano-assets.gopagoda.io/readme-headers/pulse.png)](http://nanobox.io/open-source#pulse)  
[![Build Status](https://travis-ci.org/nanopack/pulse.svg)](https://travis-ci.org/nanopack/pulse)
[![GoDoc](https://godoc.org/github.com/nanopack/pulse?status.svg)](https://godoc.org/github.com/nanopack/pulse)

# Pulse

Pulse is a stat collecting and publishing service. It serves historical stats over an http api while live stats are sent to mist for live updates.


## Usage

Simply running `pulse -s` will start pulse with the default config options.  
`pulse -h` or `pulse --help` will show more detailed usage and config options:

```
Usage:
  pulse [flags]

Flags:
  -a, --aggregate-interval int         Interval at which stats are aggregated (default 15)
  -b, --beat-interval int              Heartbeat frequency (seconds) (default 30)
  -c, --config-file string             Config file location for server
  -C, --cors-allow string              Sets the 'Access-Control-Allow-Origin' header (default "*")
  -H, --http-listen-address string     Http listen address (default "127.0.0.1:8080")
  -i, --influx-address string          InfluxDB server address (default "http://127.0.0.1:8086")
  -I, --insecure                       Run insecure (default true)
  -k, --kapacitor-address string       Kapacitor server address (http://127.0.0.1:9092)
  -l, --log-level string               Level at which to log (default "INFO")
  -m, --mist-address string            Mist server address
  -M, --mist-token string              Mist server token
  -p, --poll-interval int              Interval to request stats from clients (default 60)
  -r, --retention int                  Number of weeks to store aggregated stats (default 1)
  -s, --server                         Run as server
  -S, --server-listen-address string   Server listen address (default "127.0.0.1:3000")
  -t, --token string                   Security token (recommend placing in config file) (default "secret")
  -v, --version                        Print version info and exit
```

### Config File Options:
```json
{
  "server": true,
  "server-listen-address": "127.0.0.1:3000",
  "http-listen-address": "127.0.0.1:8080",
  "influx-address": "http://127.0.0.1:8086",
  "kapacitor-address": "http://127.0.0.1:9092",
  "insecure": true,
  "mist-address": "",
  "mist-token": "",
  "log-level": "info",
  "cors-allow": "*",
  "token": "secret",
  "poll-interval": 60,
  "aggregate-interval": 15,
  "beat-interval": 30,
  "retention": 12
}
```


## API

| Route | Description | Output |
| --- | --- | --- |
| **GET** /keys | Returns list of stats being recorded | string array |
| **GET** /tags | Returns list of filterable tags | string array |
| **GET** /latest/{stat}* | Returns latest stat (averages if multiple filters applied) | json stat object |
| **GET** /hourly/{stat}** | Returns hourly averages for stat | json array of stat objects |
| **GET** /daily/{stat}** | Returns average for stat at the same daily time | string map |

**ALERTS** (requires "kapacitor-address" to be configured)  

| Route | Description | Payload | Output |
| --- | --- | --- | --- |
| **POST** /alerts | Add a kapacitor alert | json alert object | json alert object |
| **PUT** /alerts | Update a kapacitor alert | json alert object | json alert object |
| **DELETE** /alerts/{alert} | Delete a kapacitor alert | nil | success message |

`*`: reserved query parameters is 'verb', all others act as filters  
`**`: reserved query parameters are 'backfill', 'verb', 'start', and 'stop', all others act as filters  

**note:** The API requires a token to be passed for authentication by default and is configurable at server start (`--token`). The token is passed in as a custom header: `X-AUTH-TOKEN`.  

For examples, see [the api's readme](api/README.md).


## Data Types
### Stat Object
json:
```json
{
  "time": 1465419600,
  "value": 0.75
}
```

Fields:
- **time**: Unix epoch timestamp of stat
- **value**: Numeric value of stat

### Alert Object
json:
```json
{
  "tags": {"host":"abc"},
  "metric": "cpu_used",
  "level": "crit",
  "threshold": 80,
  "duration": "30s",
  "post": "http://127.0.0.1/alert"
}
```

Fields:
- **tags**: Populates the WHERE
- **metric**: Stat to track
- **level**: Alert level (info, warn, crit)
- **threshold**: Limit that alert is triggered at
- **duration**: How far back to average (5m)
- **post**: Api to hit when alert is triggered


## Relay

A pulse relay is a service that connects to pulse and advertises stats that are available for collection. A relay implementation is available in the pulse project and can be embedded in other projects.  

For an [**example**](relay/README.md), look in the README for relay.

### TCP pulse api
The TCP api used to communicate between the pulse server and a relay is simple and is designed to be human readable and debuggable. It is newline delimited.

| Command | Description | Response |
| --- | --- | --- |
| `id {id}` | **Must** be the first command to be run, identifies the client to the server | `ok` |
| `add {name}` | Exposes a stat that can be collected by the server | `ok` |
| `remove {name}` | Removes a stat previously exposed to the server | `ok` |


### TCP relay api
| Command | Description | Response |
| --- | --- | --- |
| `get {tag,tag2}` | Request a list of stats corrosponding to the list of tags passed in | `got {tag:value}` |
| `flush` | Clear all current values from the stat collectors | `ok` |
| `override {duration} {tag:interval}` | for `duration` seconds, bump the collection interval from the default to `interval` for each `tag:interval` | `ok` |

#### Notes
- If an override is specified for a stat, and a new machine comes online and connects, that override is **NOT** honored.
- Pulse server does not actively connect to servers to have stats pushed to it, rather, it waits for stat collecting machines to connect and then requests certain stats on specific intervals.


## Contributing

Contributions to the pulse project are welcome and encouraged. Pulse is a [Nanobox](https://nanobox.io) project and contributions should follow the [Nanobox Contribution Process & Guidelines](https://docs.nanobox.io/contributing/).

#### TODO

## Licence

Mozilla Public License Version 2.0

[![open source](http://nano-assets.gopagoda.io/open-src/nanobox-open-src.png)](http://nanobox.io/open-source)


CHANGELOG.md
## v0.2.3 (Jun 20, 2017)

IMPROVEMENTS:
  - Minor improvements to logging

## v0.2.2 (Apr 14, 2017)

IMPROVEMENTS:
  - Allow integers or floats for threshold
  - Allow different comparison operators
  - Added routes for listing single and all tasks
  - Updated alerts API docs

## v0.2.1 (Feb 8, 2017)

BUG FIXES:
  - #10 Resolves client multiple heartbeats
  - #10 Resolves server's client-awareness issues
  - Fixed `beat-interval` default

IMPROVEMENTS:
  - Added dirty commit identifier

## v0.2.0 (Jan 30, 2017)

BUG FIXES:
  - Resolve connection timeout issue (for real this time)

IMPROVEMENTS:
  - Handle errors rather than panic

## v0.1.0 (Jan 3, 2017)

BUG FIXES:
  - Resolve connection timeout issue
  - Made alert ids unique

IMPROVEMENTS:
  - Added a better way to read from the connection so disconnects are detected
  - Created meaningful alert id and message
  - Alert on statechange only
  - Tests

## Previous (Jan 3, 2017)

This change log began with version 0.1.0. Any prior changes can be seen by viewing
the commit history.


main.go
// Pulse is a stat collecting and publishing service. It serves historical
// stats over an http api while live stats are sent to mist for live updates.
//
// Usage
//
// To start pulse as a server, simply run (with influx running locally):
//
//  pulse -s
//
// For more specific usage information, refer to the help doc `pulse -h`:
//  Usage:
//    pulse [flags]
//
//  Flags:
//    -a, --aggregate-interval int         Interval at which stats are aggregated (default 15)
//    -c, --config-file string              Config file location for server
//    -C, --cors-allow string              Sets the 'Access-Control-Allow-Origin' header (default "*")
//    -H, --http-listen-address string     Http listen address (default "127.0.0.1:8080")
//    -i, --influx-address string          InfluxDB server address (default "http://127.0.0.1:8086")
//    -I, --insecure                       Run insecure (default true)
//    -k, --kapacitor-address string       Kapacitor server address (http://127.0.0.1:9092)
//    -l, --log-level string               Level at which to log (default "INFO")
//    -m, --mist-address string            Mist server address
//    -M, --mist-token string              Mist server token
//    -p, --poll-interval int              Interval to request stats from clients (default 60)
//    -s, --server                         Run as server
//    -S, --server-listen-address string   Server listen address (default "127.0.0.1:3000")
//    -t, --token string                   Security token (recommend placing in config file) (default "secret")
//    -v, --version                        Print version info and exit
//
package main

import (
	"fmt"
	"time"

	"github.com/jcelliott/lumber"
	mist "github.com/nanopack/mist/clients"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"

	"github.com/nanopack/pulse/api"
	"github.com/nanopack/pulse/influx"
	"github.com/nanopack/pulse/kapacitor"
	"github.com/nanopack/pulse/plexer"
	pulse "github.com/nanopack/pulse/server"
)

var (
	// for clarity
	httpAddress       = "127.0.0.1:8080"
	serverAddress     = "127.0.0.1:3000"
	influxAddress     = "http://127.0.0.1:8086"
	kapacitorAddress  = "" // "http://127.0.0.1:9092"
	mistAddress       = ""
	mistToken         = ""
	logLevel          = "info"
	corsAllow         = "*"
	server            = false
	insecure          = true
	token             = "secret"
	pollInterval      = 60
	beatInterval      = 30 // heartbeat frequency (seconds)
	aggregateInterval = 15
	retention         = 1

	configFile = ""
	version    = false

	// Pulse is the pulse cli
	Pulse = &cobra.Command{
		Use:   "pulse",
		Short: "pulse is a stat collecting and publishing service",
		Long:  ``,

		PersistentPreRunE: readConfig,
		PreRunE:           preFlight,
		RunE:              startPulse,
		SilenceErrors:     true,
		SilenceUsage:      true,
	}

	// to be populated by go linker
	tag    string
	commit string
)

func init() {
	Pulse.Flags().StringP("http-listen-address", "H", httpAddress, "Http listen address")
	viper.BindPFlag("http-listen-address", Pulse.Flags().Lookup("http-listen-address"))
	Pulse.Flags().StringP("server-listen-address", "S", serverAddress, "Server listen address")
	viper.BindPFlag("server-listen-address", Pulse.Flags().Lookup("server-listen-address"))
	Pulse.Flags().StringP("influx-address", "i", influxAddress, "InfluxDB server address")
	viper.BindPFlag("influx-address", Pulse.Flags().Lookup("influx-address"))
	Pulse.Flags().StringP("kapacitor-address", "k", kapacitorAddress, "Kapacitor server address (http://127.0.0.1:9092)")
	viper.BindPFlag("kapacitor-address", Pulse.Flags().Lookup("kapacitor-address"))
	Pulse.Flags().StringP("mist-address", "m", mistAddress, "Mist server address")
	viper.BindPFlag("mist-address", Pulse.Flags().Lookup("mist-address"))
	Pulse.Flags().StringP("mist-token", "M", mistToken, "Mist server token")
	viper.BindPFlag("mist-token", Pulse.Flags().Lookup("mist-token"))
	Pulse.Flags().StringP("log-level", "l", logLevel, "Level at which to log")
	viper.BindPFlag("log-level", Pulse.Flags().Lookup("log-level"))
	Pulse.Flags().StringP("cors-allow", "C", corsAllow, "Sets the 'Access-Control-Allow-Origin' header")
	viper.BindPFlag("cors-allow", Pulse.Flags().Lookup("cors-allow"))
	Pulse.Flags().BoolP("server", "s", server, "Run as server")
	viper.BindPFlag("server", Pulse.Flags().Lookup("server"))
	Pulse.Flags().BoolP("insecure", "I", insecure, "Run insecure")
	viper.BindPFlag("insecure", Pulse.Flags().Lookup("insecure"))

	Pulse.Flags().StringP("token", "t", token, "Security token (recommend placing in config file)")
	viper.BindPFlag("token", Pulse.Flags().Lookup("token"))
	Pulse.Flags().IntP("poll-interval", "p", pollInterval, "Interval to request stats from clients")
	viper.BindPFlag("poll-interval", Pulse.Flags().Lookup("poll-interval"))
	Pulse.Flags().IntP("aggregate-interval", "a", aggregateInterval, "Interval at which stats are aggregated")
	viper.BindPFlag("aggregate-interval", Pulse.Flags().Lookup("aggregate-interval"))
	Pulse.Flags().IntP("retention", "r", retention, "Number of weeks to store aggregated stats")
	viper.BindPFlag("retention", Pulse.Flags().Lookup("retention"))
	Pulse.Flags().IntP("beat-interval", "b", beatInterval, "Heartbeat frequency (seconds)")
	viper.BindPFlag("beat-interval", Pulse.Flags().Lookup("beat-interval"))

	Pulse.Flags().StringVarP(&configFile, "config-file", "c", configFile, "Config file location for server")
	Pulse.Flags().BoolVarP(&version, "version", "v", version, "Print version info and exit")

	lumber.Level(lumber.LvlInt(viper.GetString("log-level")))
}

func preFlight(ccmd *cobra.Command, args []string) error {
	if version {
		fmt.Printf("pulse %s (%s)\n", tag, commit)
		return fmt.Errorf("") // no error, just exit
	}

	if !viper.GetBool("server") {
		ccmd.HelpFunc()(ccmd, args)
		return fmt.Errorf("") // no error, just exit
	}

	return nil
}

func readConfig(ccmd *cobra.Command, args []string) error {
	if configFile != "" {
		viper.SetConfigFile(configFile)
		err := viper.ReadInConfig()
		if err != nil {
			return fmt.Errorf("Failed to read config - %s", err)
		}
	}

	// validate retention
	if viper.GetInt("retention") < 1 {
		fmt.Println("Bad value for retention, resetting to 1")
		viper.Set("retention", 1)
	}

	// validate beat-interval
	if viper.GetInt("beat-interval") < 1 {
		fmt.Println("Bad value for beat-interval, resetting to 30")
		viper.Set("beat-interval", 30)
	}

	return nil
}

func main() {
	// print errors
	err := Pulse.Execute()
	if err != nil && err.Error() != "" {
		fmt.Println(err)
	}
}

func startPulse(ccmd *cobra.Command, args []string) error {
	// re-initialize logger
	lumber.Level(lumber.LvlInt(viper.GetString("log-level")))

	plex := plexer.NewPlexer()

	if viper.GetString("mist-address") != "" {
		mist, err := mist.New(viper.GetString("mist-address"), viper.GetString("mist-token"))
		if err != nil {
			return fmt.Errorf("Mist failed to start - %s", err)
		}
		plex.AddObserver("mist", mist.Publish)
		defer mist.Close()
	}

	plex.AddBatcher("influx", influx.Insert)

	err := pulse.Listen(viper.GetString("server-listen-address"), plex.Publish)
	if err != nil {
		return fmt.Errorf("Pulse failed to start - %s", err)
	}
	// begin polling the connected servers
	pollSec := viper.GetInt("poll-interval")
	if pollSec == 0 {
		pollSec = 60
	}
	go pulse.StartPolling(nil, nil, time.Duration(pollSec)*time.Second, nil)

	queries := []string{
		"CREATE DATABASE statistics",
		"CREATE RETENTION POLICY one_day ON statistics DURATION 8h REPLICATION 1 DEFAULT",
		fmt.Sprintf("CREATE RETENTION POLICY one_week ON statistics DURATION %dw REPLICATION 1", viper.GetInt("retention")), // todo: ALTER as well?
	}

	for _, query := range queries {
		_, err := influx.Query(query)
		if err != nil {
			return fmt.Errorf("Failed to query influx - %s", err)
		}
	}

	go influx.KeepContinuousQueriesUpToDate()

	if viper.GetString("kapacitor-address") != "" {
		err := kapacitor.Init()
		if err != nil {
			return fmt.Errorf("Kapacitor failed to start - %s", err)
		}
	}

	err = api.Start()
	if err != nil {
		return fmt.Errorf("Api failed to start - %s", err)
	}

	return nil
}
